### PR TITLE
Standalone TPCDS tester

### PR DESCRIPTION
**Pull request by ileshko**
Simplify running TPCDS benchmarks.

### CODE REVIEW

**Comment by npoggi** [comment]
@ileshko thanks for the PR, do you have some sample output and updated instructions for the comparison feature? will test the PR to see if it doesn't break existing benchmarks, also adding @juliuszsompolski as a reviewer.  About bundling dsdgen, not sure if we can do that in this repo, can you ask?

File: src/main/scala/com/databricks/spark/sql/perf/tpcds/DsdgenResourceExtractor.scala [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      throw new Exception(s"No dsdgen binary for platform: ${System.getProperty("os.name")}")
+    }
+
+    val tempDirPath = Files.createDirectory(Paths.get(dirName))
+    val tempDireFile = new File(tempDirPath.toString)
+
+    emitDsdgenResourceDirectory(resourceDir, tempDirPath.toString)
+    tempDirPath.resolve("dsdgen").toFile.setExecutable(true)
+
+    // If this file is already present (e.g. created by different thread), ignore
```
Comment (juliuszsompolski) [inline src/main/scala/com/databricks/spark/sql/perf/tpcds/DsdgenResourceExtractor.scala]: Maybe wrap this whole function in an if (alreadyExists) to do less racing between threads.

File: src/main/scala/com/databricks/spark/sql/perf/BenchmarkCompare.scala [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      .getOrCreate()
+
+    RunBenchmark.compareResults(
+      spark.sqlContext,
+      conf.baseline,
+      conf.target,
+      conf.resultsLocation,
+      conf.outputLocation)
+  }
+}
```
Comment (juliuszsompolski) [inline src/main/scala/com/databricks/spark/sql/perf/BenchmarkCompare.scala]: nit: no newline

File: src/main/scala/com/databricks/spark/sql/perf/Query.scala [hunk 3]
```diff
@@ -38,18 +38,16 @@ class Query(
...
-           |== Query: $name ==
-           | Can't be analyzed: $e
-           |
-           | $description
-         """.stripMargin
+        s"""|== Query: $name ==
+            | Can't be analyzed: $e
+            |
+            | $description
+          """.stripMargin
```
Comment (juliuszsompolski) [inline src/main/scala/com/databricks/spark/sql/perf/Query.scala]: nit: all changes in this file seem to just be an indent change. revert.

File: src/main/scala/com/databricks/spark/sql/perf/Benchmark.scala [hunk 4]
```diff
@@ -297,13 +297,13 @@ object Benchmark {
...
       currentConfiguration: BenchmarkConfiguration,
-      forkThread: Boolean = true) {
+      forkThread: Boolean = true) extends Logging {
     val currentResults = new collection.mutable.ArrayBuffer[BenchmarkResult]()
     val currentRuns = new collection.mutable.ArrayBuffer[ExperimentRun]()
     val currentMessages = new collection.mutable.ArrayBuffer[String]()
 
     def logMessage(msg: String) = {
-      println(msg)
+      logger.info(msg)
```
Comment (juliuszsompolski) [inline src/main/scala/com/databricks/spark/sql/perf/Benchmark.scala]: Hmm... when running in Databricks, having it as a println helps me to see what's going on with messages appearing printed out straight under the notebook cell, and not having to look for them in the Driver logs, which will be cluttered with other stuff.

File: src/main/scala/com/databricks/spark/sql/perf/tpcds/DsdgenResourceExtractor.scala [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import org.apache.commons.io.IOUtils
+import org.apache.commons.lang3.SystemUtils
+
+/**
+ * Extracts the platform-specific executable from the resource stream.
+ */
+object DsdgenResourceExtractor {
+  private val dsdgenBinaryResources = Seq(
+    "dsdgen",
+    "tpcds.idx"
```
Comment (juliuszsompolski) [inline src/main/scala/com/databricks/spark/sql/perf/tpcds/DsdgenResourceExtractor.scala]: What's that?

File: src/main/scala/com/databricks/spark/sql/perf/tpcds/TPCDS_Standalone.scala [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+      resultLocation = conf.outputLocation,
+      tags = Map(
+        "runtype" -> "benchmark",
+        "database" -> generateDbName(conf),
+        "scale_factor" -> conf.scaleFactor.toString))
+
+    logger.info(experiment.toString)
+    experiment.waitForFinish(conf.timeoutHours * 60 * 60)
+
+    spark.sqlContext.setConf("spark.sql.shuffle.partitions", "1")
```
Comment (juliuszsompolski) [inline src/main/scala/com/databricks/spark/sql/perf/tpcds/TPCDS_Standalone.scala]: better explicitly .repartition(1) or .coalesce(1) where you need it? (and you seem to already be doing that)

File: src/main/scala/com/databricks/spark/sql/perf/tpcds/DsdgenResourceExtractor.scala [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    val resourceDir = if (SystemUtils.IS_OS_LINUX) {
+      "dsdgen-kit/linux"
+    } else if (SystemUtils.IS_OS_MAC_OSX) {
+      "dsdgen-kit/osx"
+    } else {
+      throw new Exception(s"No dsdgen binary for platform: ${System.getProperty("os.name")}")
+    }
+
+    val tempDirPath = Files.createDirectory(Paths.get(dirName))
+    val tempDireFile = new File(tempDirPath.toString)
```
Comment (juliuszsompolski) [inline src/main/scala/com/databricks/spark/sql/perf/tpcds/DsdgenResourceExtractor.scala]: nit: Dire -> Dir

File: src/main/scala/com/databricks/spark/sql/perf/RunBenchmark.scala [hunk 8]
```diff
@@ -101,39 +100,102 @@ object RunBenchmark {
...
+      targetTimestamp: Long,
+      resultsLocation: String,
+      outputLocation: String): Unit = {
+
+    import sqlContext.implicits._
+    val baselineTime = when($"timestamp" === baselineTimestamp, $"executionTime").otherwise(null)
+    val targetTime = when($"timestamp" === targetTimestamp, $"executionTime").otherwise(null)
+
+    val comparison = sqlContext.read.json(resultsLocation)
+        .coalesce(1)
```
Comment (juliuszsompolski) [inline src/main/scala/com/databricks/spark/sql/perf/RunBenchmark.scala]: why is it needed here?
It's going to get uncoalesced by the agg later (but you do another coalesce before writing anyway).

File: src/main/scala/com/databricks/spark/sql/perf/RunBenchmark.scala [hunk 9]
```diff
@@ -101,39 +100,102 @@ object RunBenchmark {
...
+        .groupBy("name")
+        .agg(
+          avg(baselineTime) as 'baselineTimeMs,
+          avg(targetTime) as 'targetTimeMs,
+          stddev(baselineTime) as 'stddev)
+        .withColumn(
+          "percentChange", ($"baselineTimeMs" - $"targetTimeMs") / $"baselineTimeMs" * 100)
+        .filter('targetTimeMs.isNotNull)
+
+    comparison.show(comparison.count.toInt, truncate = false)
```
Comment (juliuszsompolski) [inline src/main/scala/com/databricks/spark/sql/perf/RunBenchmark.scala]: nit: don't want some order by (e.g. by query name)  before showing / writing it?

File: src/main/scala/com/databricks/spark/sql/perf/tpcds/TPCDS_Standalone.scala [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+
+  def generateDbName(conf: TpcdsStandaloneConfig): String = {
+    s"tpcds_sf${conf.scaleFactor}" +
+      s"""_${if (conf.useDecimal) "with" else "no"}decimal""" +
+      s"""_${if (conf.useDate) "with" else "no"}date""" +
+      s"""_${if (conf.filterNull) "no" else "with"}nulls"""
+  }
+
+  def generateDataset(spark: SparkSession, conf: TpcdsStandaloneConfig): Unit = {
+      // Limit the memory used by parquet writer
```
Comment (juliuszsompolski) [inline src/main/scala/com/databricks/spark/sql/perf/tpcds/TPCDS_Standalone.scala]: nit: indent

File: src/main/scala/com/databricks/spark/sql/perf/tpcds/TPCDS_Standalone.scala [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+  scaleFactor: Int = 10,
+  iterations: Int = 1,
+  regenerateDataset: Boolean = false,
+  useDecimal: Boolean = true,
+  useDate: Boolean = true,
+  filterNull: Boolean = false,
+  shuffle: Boolean = true,
+  timeoutHours: Int = 60,
+  format: String = "parquet",
+  shufflePartitions: Int = 12,
```
Comment (juliuszsompolski) [inline src/main/scala/com/databricks/spark/sql/perf/tpcds/TPCDS_Standalone.scala]: Why 12?
Seems like a bad idea to set it that low.
Keeping the Spark default of 200 seems ok to me.

File: src/main/scala/com/databricks/spark/sql/perf/tpcds/TPCDS_Standalone.scala [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+          partitionTables = true,
+          clusterByPartitionColumns = conf.shuffle,
+          filterOutNullPartitionValues = conf.filterNull,
+          tableFilter = t,
+          numPartitions = dsdgen_nonpartitioned)
+    }}
+    logger.info("Done generating non partitioned tables.")
+
+    // leave the biggest/potentially hardest tables to be generated last.
+    val dsdgen_partitioned = 10000 // recommended for SF10000+.
```
Comment (juliuszsompolski) [inline src/main/scala/com/databricks/spark/sql/perf/tpcds/TPCDS_Standalone.scala]: Hm. maybe make it SF dependent?
roughly = SF, with a MIN of 10 maybe?

File: src/main/scala/com/databricks/spark/sql/perf/tpcds/TPCDS_Standalone.scala [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+    tables.createExternalTables(
+      rootDir,
+      conf.format,
+      databaseName,
+      overwrite = true,
+      discoverPartitions = true)
+    logger.info("External table generation complete")
+
+    // For CBO
+    logger.info("Generating CBO data")
```
Comment (juliuszsompolski) [inline src/main/scala/com/databricks/spark/sql/perf/tpcds/TPCDS_Standalone.scala]: nit: data -> statistics

File: src/main/scala/com/databricks/spark/sql/perf/BenchmarkCompare.scala [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+      .appName(getClass.getName)
+      .getOrCreate()
+
+    RunBenchmark.compareResults(
+      spark.sqlContext,
+      conf.baseline,
+      conf.target,
+      conf.resultsLocation,
+      conf.outputLocation)
+  }
```
Comment (liancheng) [inline src/main/scala/com/databricks/spark/sql/perf/BenchmarkCompare.scala]: Missing `spark.stop()` at the end.

File: src/main/scala/com/databricks/spark/sql/perf/tpcds/DsdgenResourceExtractor.scala (hunk 5 above)
Comment (ileshko) [reply src/main/scala/com/databricks/spark/sql/perf/tpcds/DsdgenResourceExtractor.scala]: `tpcds.idx` is a binary file that is produced along with `dsdgen` upon compilation. It is a binary file that defines the random distributions of the data in the TPC-DS source tables. This file is read by `dsdgen` at the TPC-DS dataset generation time.

**Comment by npoggi** [comment]
@ileshko @liancheng should we get this PR to a mergeable status? Or maybe branch it?

**Comment by liancheng** [comment]
@npoggi Yes, I'm going to do that, sorry for the delay.

**Comment by shivajid** [comment]
Can this be merged into the mainline code?

**Comment by npoggi** [comment]
@shivajid there are still comments unaddressed in the PR

**Pull request closed without merging**